benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,223 @@
+
+struct GPUAtomic {
+    using policy = RAJA::policy::hip::hip_atomic;
+};
+
+template<typename AtomicType>
+void AllocateAtomicDevice(AtomicType** atomic, int array_length) {
+    hipMalloc((void **)atomic, len_array * sizeof(AtomicType));
+    hipMemset(*atomic, 0, len_array * sizeof(AtomicType));
+}
+
+template<typename AtomicType>
+void DeallocateDeviceAtomic(AtomicType* atomic) {
+    hipFree((void *)atomic);
+}
+
+#endif
+
+#include "desul/atomics.hpp"
+#include "RAJA/util/Timer.hpp"
+
+#define N 1000000000
+#define INDENT "  "
+using raja_default_desul_order = desul::MemoryOrderRelaxed;
+using raja_default_desul_scope = desul::MemoryScopeDevice;
+
+// Desul atomics have a different signature than RAJA's built in ops. The following code provides some
+// helper function templates so that they can be called using the same signature in timing code.
+
+// Struct holding Desul atomic signature typedef
+template<typename AtomicType>
+struct DesulAtomicSignature {
+    using signature = AtomicType(*)(AtomicType*, const AtomicType, raja_default_desul_order, raja_default_desul_scope);
+};
+
+// Struct holding RAJA atomic signature typedef
+template<typename AtomicType>
+struct RajaAtomicSignature {
+    using signature = AtomicType(*)(AtomicType*, const AtomicType);
+};
+
+/// RAJA::atomicAdd is overloaded and has an ambiguous type so it can't be passed as a template parameter.
+/// The following wrappers disambiguate the call and provide a signature comaptible with the DESUL
+/// wrapper.
+template<typename AtomicType, typename Policy>
+RAJA_HOST_DEVICE AtomicType AtomicAdd(AtomicType* acc, const AtomicType val) {
+    return RAJA::atomicAdd(Policy {}, acc, val);
+}
+
+template<typename AtomicType, typename Policy>
+RAJA_HOST_DEVICE AtomicType AtomicMax(AtomicType* acc, const AtomicType val) {
+    return RAJA::atomicMax(Policy {}, acc, val);
+}
+
+/// Function template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+template<typename T, typename Policy, typename DesulAtomicSignature<T>::signature AtomicImpl>
+RAJA_HOST_DEVICE T atomicWrapperDesul(T * acc, T value) {
+    return AtomicImpl(acc, value, raja_default_desul_order{},
+                    raja_default_desul_scope{});
+}
+
+//template<typename T, typename RajaAtomicSignature<T>::signature atomic>
+//class IsDesul : public std::false_type {};
+//
+//template<typename T, typename Policy, typename DesulAtomicSignature<T>::signature AtomicImpl>
+//class IsDesul<atomicWrapperDesul<T, Policy, AtomicImpl>> : public std::true_type {};
+
+
+template<typename AtomicType, typename Policy>
+std::string GetImplName (typename DesulAtomicSignature<AtomicType>::signature) {
+    return "Desul atomic";
+}
+
+template <class ExecPolicy, typename AtomicType, typename RajaAtomicSignature<AtomicType>::signature AtomicImpl, bool test_array = false>
+void TimeAtomicOp(int num_iterations = 2, int array_size = 100) {
+    RAJA::Timer timer;
+
+    for (int i = 0; i < num_iterations; ++i) {
```

Comment:
It might be good to not time the first run of the kernel. Then time multiple iterations of the loop, running asynchronously for gpus, instead of timing each loop individually.

Reply:
should I move the memory allocations / deallocations out of the timing loop in this case?

Reply:
Ya, you don't want to time anything extra.